src/Microsoft.Azure.Batch.SoftwareEntitlement.Common/Option.cs

```diff
@@ -0,0 +1,302 @@
+using System;
+
+namespace Microsoft.Azure.Batch.SoftwareEntitlement.Common
+{
+    /// <summary>
+    /// Represents an optional result.
+    /// </summary>
+    /// <typeparam name="T">The result type.</typeparam>
+    public interface IOption<T> : IEquatable<IOption<T>>
+    {
+        /// <summary>
+        /// Perform one of two functions based on this option
+        /// </summary>
+        /// <typeparam name="R">Return type</typeparam>
+        /// <param name="none">Function to perform for None{T}."/></param>
+        /// <param name="some">Function to perform for Some{T}.</param>
+        /// <returns></returns>
+        R Match<R>(Func<R> none, Func<T, R> some);
+
+        /// <summary>
+        /// Perform one of two actions based on this option
+        /// </summary>
+        /// <param name="none">Action to perform for None{T}."/></param>
+        /// <param name="some">Action to perform for Some{T}.</param>
+        void Match(Action none, Action<T> some);
+
+        /// <summary>
+        /// Act when this option is None{T}
+        /// </summary>
+        /// <param name="action">Action to invoke.</param>
+        /// <returns>This option, for chaining.</returns>
+        IOption<T> WhenNone(Action action);
+
+        /// <summary>
+        /// Act when this option is Some{T}
+        /// </summary>
+        /// <param name="action">Action to invoke.</param>
+        /// <returns>This option, for chaining.</returns>
+        IOption<T> WhenSome(Action<T> action);
+
+        /// <summary>
+        /// Return a default value if no value is available
+        /// </summary>
+        /// <param name="defaultValue">Default value to return</param>
+        /// <returns>
+        /// Wrapped value (if <see cref="Some{T}"/>) or default value (if <see cref="None{T}"/>).
+        /// </returns>
+        T OrDefault(T defaultValue);
+    }
+
+    /// <summary>
+    /// Static factory methods for option values
+    /// </summary>
+    public static class Option
+    {
+        /// <summary>
+        /// Create a Some{T} for a given value of T
+        /// </summary>
+        /// <typeparam name="T">Type of value to wrap.</typeparam>
+        /// <param name="value">Value to wrap.</param>
+        /// <returns>New instance.</returns>
+        public static IOption<T> Some<T>(T value)
+        {
+            return new Some<T>(value);
+        }
+
+        /// <summary>
+        /// Create a None{T} for a given type
+        /// </summary>
+        /// <typeparam name="T">Type of value we don't have.</typeparam>
+        /// <returns>New instance.</returns>
+        public static IOption<T> None<T>()
```

Comment:
consider having this as a property `public static IOption<T> None { get; } = new None<T>();`

Reply:
Wrapper type isn't generic, so I'd have to create a cache dictionary to do this. Maybe later. #wontfix